### ISSUE TITLE
No download activity when network interface is set

### ISSUE THREAD

**Issue by evsh** [arvidn/libtorrent#3034]
**Please provide the following information**

libtorrent version (or branch): master

platform/architecture: Gentoo Linux, amd64

compiler and compiler version: GCC 8

Attempting to download the Arch Linux ISO using the example python client:
```
./client.py -i tun0 -o tun0 -s /tmp 'magnet:?xt=urn:btih:c1e797e823900db479b3ec2fb3bd18fe44c89f30&dn=archlinux-2018.05.01-x86_64.iso&tr=udp://tracker.archlinux.org:6969&tr=http://tracker.archlinux.org:6969/announce'
```

Console output contains the following error messages:
```
listening on 10.25.0.5:6881 (device: tun0) failed: [sock_bind_to_device] [TCP] Operation not permitted
listening on [fe80::blah-blah-blah%tun0]:6881 (device: tun0) failed: [sock_bind_to_device] [TCP] Operation not permitted
listening on [fe80::blah-blah-blah%tun0]:6881 (device: tun0) failed: [sock_bind_to_device] [TCP] Operation not permitted
```
Downloading starts only after removing both `-o` and `-i` options.

**Comment by evsh** [comment]
Let me explain myself. Sure, there are those who want this SO_BINDTODEVICE behaviour, but there are other users, and I believe those other outnumber those who run torrent daemons, who simply want their torrent traffic does not escape an interface. These options (listen_interface and outgoing_interface) were providing the way to achieve that in RC_1_1. Also note, please, that at the time of #2969 only the outgoing_ibterface option was blocking download activity when used by a regular user, while this report states that the listen_interface alone does block traffic too. Thus I see that as a clear 1.1 to 1.2 regression.

**Comment by arvidn** [comment]
it does seem reasonable that for the listen sockets `SO_BINDTODEVICE` could be best-effort. Accepting incoming connections honours the bind interface. For outgoing sockets though, it must be set, otherwise the system wide routing table will be used, and what the socket was bound to is ignored

**Comment by arvidn** [comment]
As far as I can tell from looking at the [code](https://github.com/arvidn/libtorrent/blob/master/include/libtorrent/enum_net.hpp#L134), bind-to-device *is* already best effort.

Notice how if `set_option()` fails, the error is cleared and we fall back to enumerating the existing interfaces, looking for a match and use regular `bind()` on that address.

The case were it would fail would be if the device name cannot be found (which seems reasonable, as that would probably be a type in the settings).

Do you know where the error/failure you're seeing is coming from?

**Comment by arvidn** [comment]
ah, I suppose [this](https://github.com/arvidn/libtorrent/blob/master/src/session_impl.cpp#L1440) is the problem. it should probably use the utility function template from `enum_net.hpp`

**Comment by arvidn** [comment]
This issue does not affect `RC_1_1`, right?

**Comment by evsh** [comment]
Yes, 1.1.7 works fine. Did not test with the branch head.

**Comment by evsh** [comment]
Sorry for the late reply. Here are the stacktraces, please:
`listen_failed_alert`:
```
#0 libtorrent::listen_failed_alert::listen_failed_alert() at /home/eugene/develop/qbt/libtorrent/src/alert.cpp:933
#1 libtorrent::heterogeneous_queue<libtorrent::alert>::emplace_back<libtorrent::listen_failed_alert, libtorrent::aux::stack_allocator&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>&, libtorrent::operation_t&, boost::system::error_code&, libtorrent::socket_type_t const&>() at /home/eugene/develop/qbt/libtorrent/include/libtorrent/heterogeneous_queue.hpp:110
#2 libtorrent::alert_manager::emplace_alert<libtorrent::listen_failed_alert, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, boost::asio::ip::basic_endpoint<boost::asio::ip::tcp>&, libtorrent::operation_t&, boost::system::error_code&, libtorrent::socket_type_t const&>() at /home/eugene/develop/qbt/libtorrent/include/libtorrent/alert_manager.hpp:90
#3 libtorrent::aux::session_impl::setup_listener() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:1454
#4 libtorrent::aux::session_impl::reopen_listen_sockets() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:1888
#5 libtorrent::aux::session_impl::init() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:589
#6 libtorrent::aux::session_impl::wrap<void (libtorrent::aux::session_impl::*)()>() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:458
#7 libtorrent::aux::session_impl::<lambda()>::operator()(void) const() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:542
#8 boost::asio::asio_handler_invoke<libtorrent::aux::session_impl::start_session()::<lambda()> >(libtorrent::aux::session_impl::<lambda()> &, ...)() at /usr/include/boost/asio/handler_invoke_hook.hpp:69
#9 boost_asio_handler_invoke_helpers::invoke<libtorrent::aux::session_impl::start_session()::<lambda()>, libtorrent::aux::session_impl::start_session()::<lambda()> >(libtorrent::aux::session_impl::<lambda()> &, libtorrent::aux::session_impl::<lambda()> &)() at /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37
#10 boost::asio::detail::handler_work<libtorrent::aux::session_impl::start_session()::<lambda()>, boost::asio::system_executor>::complete<libtorrent::aux::session_impl::start_session()::<lambda()> >(libtorrent::aux::session_impl::<lambda()> &, libtorrent::aux::session_impl::<lambda()> &)() at /usr/include/boost/asio/detail/handler_work.hpp:82
#11 boost::asio::detail::completion_handler<libtorrent::aux::session_impl::start_session()::<lambda()> >::do_complete(void *, boost::asio::detail::operation *, const boost::system::error_code &, std::size_t)() at /usr/include/boost/asio/detail/completion_handler.hpp:70
#12 boost::asio::detail::scheduler_operation::complete() at /usr/include/boost/asio/detail/scheduler_operation.hpp:40
#13 boost::asio::detail::scheduler::do_run_one() at /usr/include/boost/asio/detail/impl/scheduler.ipp:401
#14 boost::asio::detail::scheduler::run() at /usr/include/boost/asio/detail/impl/scheduler.ipp:154
#15 boost::asio::io_context::run() at /usr/include/boost/asio/impl/io_context.ipp:62
#16 libtorrent::session::<lambda()>::operator()(void) const() at /home/eugene/develop/qbt/libtorrent/src/session.cpp:363
#17 std::__invoke_impl<void, libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> >(std::__invoke_other, libtorrent::session::<lambda()> &&)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/bits/invoke.h:60
#18 std::__invoke<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> >(libtorrent::session::<lambda()> &&)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/bits/invoke.h:95
#19 std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/thread:234
#20 std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> > >::operator()(void)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/thread:243
```
and for the `udp_error_alert`:

```
#0 libtorrent::udp_error_alert::udp_error_alert() at /home/eugene/develop/qbt/libtorrent/src/alert.cpp:1024
#1 libtorrent::heterogeneous_queue<libtorrent::alert>::emplace_back<libtorrent::udp_error_alert, libtorrent::aux::stack_allocator&, boost::asio::ip::basic_endpoint<boost::asio::ip::udp> const&, boost::system::error_code&>() at /home/eugene/develop/qbt/libtorrent/include/libtorrent/heterogeneous_queue.hpp:110
#2 libtorrent::alert_manager::emplace_alert<libtorrent::udp_error_alert, boost::asio::ip::basic_endpoint<boost::asio::ip::udp> const&, boost::system::error_code&>() at /home/eugene/develop/qbt/libtorrent/include/libtorrent/alert_manager.hpp:90
#3 libtorrent::aux::session_impl::reopen_outgoing_sockets() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:2058
#4 libtorrent::aux::session_impl::init() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:590
#5 libtorrent::aux::session_impl::wrap<void (libtorrent::aux::session_impl::*)()>() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:458
#6 libtorrent::aux::session_impl::<lambda()>::operator()(void) const() at /home/eugene/develop/qbt/libtorrent/src/session_impl.cpp:542
#7 boost::asio::asio_handler_invoke<libtorrent::aux::session_impl::start_session()::<lambda()> >(libtorrent::aux::session_impl::<lambda()> &, ...)() at /usr/include/boost/asio/handler_invoke_hook.hpp:69
#8 boost_asio_handler_invoke_helpers::invoke<libtorrent::aux::session_impl::start_session()::<lambda()>, libtorrent::aux::session_impl::start_session()::<lambda()> >(libtorrent::aux::session_impl::<lambda()> &, libtorrent::aux::session_impl::<lambda()> &)() at /usr/include/boost/asio/detail/handler_invoke_helpers.hpp:37
#9 boost::asio::detail::handler_work<libtorrent::aux::session_impl::start_session()::<lambda()>, boost::asio::system_executor>::complete<libtorrent::aux::session_impl::start_session()::<lambda()> >(libtorrent::aux::session_impl::<lambda()> &, libtorrent::aux::session_impl::<lambda()> &)() at /usr/include/boost/asio/detail/handler_work.hpp:82
#10 boost::asio::detail::completion_handler<libtorrent::aux::session_impl::start_session()::<lambda()> >::do_complete(void *, boost::asio::detail::operation *, const boost::system::error_code &, std::size_t)() at /usr/include/boost/asio/detail/completion_handler.hpp:70
#11 boost::asio::detail::scheduler_operation::complete() at /usr/include/boost/asio/detail/scheduler_operation.hpp:40
#12 boost::asio::detail::scheduler::do_run_one() at /usr/include/boost/asio/detail/impl/scheduler.ipp:401
#13 boost::asio::detail::scheduler::run() at /usr/include/boost/asio/detail/impl/scheduler.ipp:154
#14 boost::asio::io_context::run() at /usr/include/boost/asio/impl/io_context.ipp:62
#15 libtorrent::session::<lambda()>::operator()(void) const() at /home/eugene/develop/qbt/libtorrent/src/session.cpp:363
#16 std::__invoke_impl<void, libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> >(std::__invoke_other, libtorrent::session::<lambda()> &&)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/bits/invoke.h:60
#17 std::__invoke<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> >(libtorrent::session::<lambda()> &&)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/bits/invoke.h:95
#18 std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> > >::_M_invoke<0>(std::_Index_tuple<0>)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/thread:234
#19 std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> > >::operator()(void)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/thread:243
#20 std::thread::_State_impl<std::thread::_Invoker<std::tuple<libtorrent::session::start(libtorrent::session_params, libtorrent::io_service*)::<lambda()> > > >::_M_run(void)() at /usr/lib64/gcc/x86_64-pc-linux-gnu/8.1.0/include/g++-v8/thread:186
```

**Comment by arvidn** [comment]
@evsh does this look reasonable? https://github.com/arvidn/libtorrent/pull/3043

**Comment by evsh** [comment]
Fixed by #3043.
